Eliminate redundant allocation in to!string for types with toString

deveshidwivedi opened the pull request:
This PR fixes the issue where `to!string` creates an unnecessary allocation when used on types that already have a `toString()` method. By adding a specialized implementation for these types, we directly use the string returned by `toString()` instead of creating a new allocation. This improves performance and reduces memory usage in common type conversion scenarios.

0xEAB commented:
Please note that the commit message is wrong. This is not Bugzilla isuue no.10560, but issue #10560 here on GitHub.

0xEAB commented:
Please add a unittest similar to the assertion outlined in the bug report.

deveshidwivedi commented on std/conv.d:
```diff
@@ -1013,7 +1013,17 @@ if (!(is(S : T) &&
     !isInfinite!S && isExactSomeString!T)
 {
-    static if (isExactSomeString!S && value[0].sizeof == ElementEncodingType!T.sizeof)
+    static if (is(typeof(T.init.toString())) && __traits(compiles, T.init.toString()) &&
```
I initially thought checking both would handle any edge cases that might occur but it was not required

pbackus commented on std/conv.d:
```diff
@@ -1013,7 +1013,13 @@ if (!(is(S : T) &&
-    static if (isExactSomeString!S && value[0].sizeof == ElementEncodingType!T.sizeof)
+    static if (is(typeof(S.init.toString())) &&
+               is(typeof(S.init.toString()) == string) &&
+               !is(S == class))
```
Is there a reason this doesn't work for classes? If so, might be worth adding a comment—it's not obvious just from looking at the code.

deveshidwivedi replied:
Yes, when `toString()` is called on a struct, it returns an immutable string, avoiding extra allocation. However, for classes,`toString()` is virtual and can be overridden, so it uses an `Appender` to ensure a separate allocation.

 I'll add a comment regarding this.

pbackus replied:
I don't understand why `toString` being virtual makes a difference here.

In D, `string` is an `alias` for `immutable(char)[]`. A `string` is always immutable, no matter what kind of function it's returned from.

deveshidwivedi replied:
I understand, since string is always immutable, the issue isn't about immutability itself but how `toString` is implemented and how `to!string` processes the result.
For structs, `toString()` typically returns a string directly, which it can use without extra allocation. Since `to!string` for classes uses an `Appender`, it ensures a separate allocation to prevent unintended sharing of memory across different instances. 
Please let me know if I am missing anything?

Herringway replied:
What's wrong with sharing immutable memory?

deveshidwivedi replied:
Got it, I misunderstood. Apologies.

0xEAB commented on std/conv.d:
```diff
@@ -1121,6 +1126,19 @@ if (!(is(S : T) &&
+@system unittest
+{
+    import std.conv : to;
+    import std.exception : enforce;
```
```suggestion
```

pbackus commented:
> Error: return value `value.toString()` of type `string` does not match return type `char[]`, and cannot be implicitly converted

Looks like we need an additional check to test whether the return type of `toString` matches the desired return type of `to`.

deveshidwivedi commented:
> > Error: return value `value.toString()` of type `string` does not match return type `char[]`, and cannot be implicitly converted
> 
> Looks like we need to test whether the return type of `toString` matches the desired return type of `to`.

I'll add a conditional check for this

pbackus commented:
This is the test that's failing in `unit-threaded`:

https://github.com/atilaneves/unit-threaded/blob/f06b64e8f60c2a109d6195c46299548564257e6f/tests/unit_threaded/ut/should.d#L533-L544

Looks like the problem is that we're attempting to call `toString` on a `null` class reference, and that's causing the program to crash.

To fix it, if `value` is a `null` class reference, we should return `null` instead of calling `value.toString()`.

deveshidwivedi commented:
I made three key changes to address the failing test cases:

- returning null if value is a null class reference
- added special handling for void and int return types 

the failing test cases:
```
std/variant.d(502): Error: expression (*zis).opDispatch() is void and has no value target = (zis).toString(); ^ std/variant.d(731): Error: template instance std.variant.VariantN!32LU.VariantN.handler!(SWithNoLength) error instantiating fptr = &handler!(T); ^ std/variant.d(656): instantiated from here: opAssign!(SWithNoLength) opAssign(value); ^ std/variant.d(3107): instantiated from here: __ctor!(SWithNoLength) Variant v = sWithNoLength; ^ make: *** [Makefile:406: generated/linux/release/64/unittest/std/variant.o] Error 1
 ```

- added support for struct types with alias this
[the failing test case](https://github.com/deveshidwivedi/phobos/blob/3f990a7e2527350730ade76f54d95bceea688d89/std/conv.d#L1384):
```
@safe unittest
{
    // Conversion representing class object with string
    class A
    {
        override string toString() @safe const { return "an A"; }
    }
    A a;
    assert(to!string(a) == "null");
    a = new A;
    assert(to!string(a) == "an A");

    // https://issues.dlang.org/show_bug.cgi?id=7660
    class C { override string toString() @safe const { return "C"; } }
    struct S { C c; alias c this; }
    S s; s.c = new C();
    assert(to!string(s) == "C");
}
```

pbackus commented:
Current failing test is here:

https://github.com/dlang-community/libdparse/blob/master/src/dparse/lexer.d#L2856

I'll be honest, this one is complex enough that it's not obvious to me what the problem is.

pbackus commented:
I'm reluctant to approve this for three reasons:

1. The added code is very complex. I can't tell whether or not it's correct just by looking at it.
2. Previous revisions of this PR have caused unexpected regressions in existing code, which means there's a higher-than-normal chance that the current revision will too.
3. Proper test coverage would make me less concerned about both (1) and (2). However, only one new `unittest` has been added.

pbackus commented:
As I suspected, there are several cases where this PR does not preserve the behavior of the original implementation, and would introduce regressions.

The current implementation relies on `std.format.write.formatValue` to convert `struct` and `class` values to strings. The actual implementation is in `std.format.internal.write.formatValueImpl`. Specifically, it is in these two overloads:

* [`formatValueImpl` for `class` values][1]
* [`formatValueImpl` for `struct` and `union` values][2]

In order to avoid regressions, any new code we introduce for calling `toString` needs to match the behavior of these functions exactly.

For structs, this is not terribly difficult, but for classes, it would require a large amount of code duplication. So for now, it may be a better idea to limit this fix to structs, and leave classes the way they are.

[1]: https://github.com/dlang/phobos/blob/67afde4f10d41d4168663f839e053298b306564d/std/format/internal/write.d#L2251
[2]: https://github.com/dlang/phobos/blob/67afde4f10d41d4168663f839e053298b306564d/std/format/internal/write.d#L2512

deveshidwivedi commented:
Thank you @pbackus, updated to align with `formatValueImpl` behavior

0xEAB commented on std/conv.d:
```diff
@@ -1013,7 +1013,22 @@ if (!(is(S : T) &&
+               !__traits(isDisabled, S.toString) &&
+               is(typeof(value.toString()) == string))
+    {
+            static if (is(T == char[]))
+        {
```
Nitpick: Please align the header of `static if` with the brace opening its body.

pbackus commented on std/conv.d:
```diff
@@ -1013,7 +1013,22 @@ if (!(is(S : T) &&
 {
-    static if (isExactSomeString!S && value[0].sizeof == ElementEncodingType!T.sizeof)
+    static if (is(S == struct) &&
+               __traits(hasMember, S, "toString") &&
+               isSomeFunction!(value.toString) &&
+               !__traits(isDisabled, S.toString) &&
+               is(typeof(value.toString()) == string))
```
```suggestion
               is(typeof(value.toString()) Ret : T) &&
               isSomeString!Ret
```
The condition here should be based on this one in `std.internal.format.write.hasToString`: https://github.com/dlang/phobos/blob/67afde4f10d41d4168663f839e053298b306564d/std/format/internal/write.d#L1896

pbackus commented on std/conv.d:
```diff
@@ -1013,7 +1013,22 @@ if (!(is(S : T) &&
+               !__traits(isDisabled, S.toString) &&
+               is(typeof(value.toString()) == string))
+    {
+            static if (is(T == char[]))
+        {
+            return value.toString().dup;
+        }
+        else
+        {
+            return value.toString();
+        }
```
```suggestion
        return value.toString();
```
If we fix the condition above, we will not need a separate check here.

pbackus commented:
> Assertion failed: j == 61440 && k == 3840, file runnable\exe1.c, line 1449

Failure is an ImportC test and appears to also be affecting other Phobos PRs, so it is almost certainly not related to the changes in this PR.

pbackus commented:
The test failure appears to have been introduced in https://github.com/dlang/dmd/pull/21010.

dkorpel commented:
That PR only touches dead code (the `-arm` switch isn't used yet), so I can't imagine it introducing the failure.

pbackus commented:
Well, the test itself was introduced in commit dlang/dmd@4261907, and the earliest instance I could find of that test failing was in the linked PR, which was merged as commit dlang/dmd@0312e82. So either the test itself has been (nondeterministically) broken since its introduction, or the failure was introduced in some commit between those two.

Looking through `git log` for commits that touch non-ARM backend code turns up the following suspects:

* dlang/dmd#20996
* dlang/dmd#20988
* dlang/dmd#20987

dkorpel commented:
The failure is nondeterministic, since several PRs have been green since then. I also only see it on Windows x64 with LDC 1.40 as host compiler. That suggests it's more than a simple logic bug in dmd, but something much more devious like undefined behavior.

0xEAB commented:
@dkorpel <https://github.com/dlang/dmd/issues/20985>
